metalearners/grid_search.py

```diff
@@ -0,0 +1,300 @@
+    X_test: Matrix | None
+    y_test: Vector | None
+    w_test: Vector | None
+    oos_method: OosMethod
+    scoring: Scoring | None
+    kwargs: dict
+
+
+@dataclass(frozen=True)
+class _GSResult:
+    r"""Cross Validation Result."""
+
+    metalearner: MetaLearner
+    train_scores: dict
+    test_scores: dict | None
+    fit_time: float
+    score_time: float
+
+
+def _fit_and_score(job: _FitAndScoreJob) -> _GSResult:
+    start_time = time.time()
+    job.metalearner.fit(job.X_train, job.y_train, job.w_train, **job.kwargs)
+    fit_time = time.time() - start_time
+
+    train_scores = job.metalearner.evaluate(
+        X=job.X_train,
+        y=job.y_train,
+        w=job.w_train,
+        is_oos=False,
+        scoring=job.scoring,
+    )
+    if job.X_test is not None and job.y_test is not None and job.w_test is not None:
+        test_scores = job.metalearner.evaluate(
+            X=job.X_test,
+            y=job.y_test,
+            w=job.w_test,
+            is_oos=True,
+            oos_method=job.oos_method,
+            scoring=job.scoring,
+        )
+    else:
+        test_scores = None
+    score_time = time.time() - fit_time
+    return _GSResult(
+        metalearner=job.metalearner,
+        fit_time=fit_time,
+        score_time=score_time,
+        train_scores=train_scores,
+        test_scores=test_scores,
+    )
+
+
+def _format_results(results: Sequence[_GSResult]) -> pd.DataFrame:
+    rows = []
+    for result in results:
+        row: dict[str, str | int | float] = {}
+        row["metalearner"] = result.metalearner.__class__.__name__
+        nuisance_models = (
+            set(result.metalearner.nuisance_model_specifications().keys())
+            - result.metalearner._prefitted_nuisance_models
+        )
+        treatment_models = set(
+            result.metalearner.treatment_model_specifications().keys()
+        )
+        for model_kind in nuisance_models:
+            row[model_kind] = result.metalearner.nuisance_model_factory[
+                model_kind
+            ].__name__
+            for param, value in result.metalearner.nuisance_model_params[
+                model_kind
+            ].items():
+                row[f"{model_kind}_{param}"] = value
+        for model_kind in treatment_models:
+            row[model_kind] = result.metalearner.treatment_model_factory[
+                model_kind
+            ].__name__
+            for param, value in result.metalearner.treatment_model_params[
+                model_kind
+            ].items():
+                row[f"{model_kind}_{param}"] = value
+        row["fit_time"] = result.fit_time
+        row["score_time"] = result.score_time
+        for name, value in result.train_scores.items():
+            row[f"train_{name}"] = value
+        if result.test_scores is not None:
+            for name, value in result.test_scores.items():
+                row[f"test_{name}"] = value
+        rows.append(row)
+    df = pd.DataFrame(rows)
```

Comment:
Nit: Might it make sense to -- instead of creating a DataFrame with dummy indices -- create a DataFrame where the "inputs" (i.e. model factory, hyperparameters) build a multi-index and the "outputs" (i.e. durations and scores) are the columns? On the one hand this might be a more canonical use of pandas[*]. On the other hand, that might make aggregation queries over different constellations a little more handy.

[*] I once read that if one uses a dummy index for a pandas DataFrame, one probably should probably use a different data structures. I thought that this was an interesting thought, but of course, ymmv.

Reply:
> [*] I once read that if one uses a dummy index for a pandas DataFrame, one probably should probably use a different data structures. I thought that this was an interesting thought, but of course, ymmv.

I never though of it but it makes sense. I did this indexing and added some lines to the tests:
[991b2f1](https://github.com/Quantco/metalearners/pull/9/commits/991b2f1696d0216339dffe12bb481d1ffce5d609)

Reply:
Cool, thanks :)